dashboard/pages/api/metric.ts

```diff
@@ -17,17 +17,147 @@
+  const map = new Map<string, number>()
+  for (const item of backPressures) {
+    const key = `${item.fragmentId}-${item.downstreamFragmentId}`
+    if (mapValue.has(key) && mapNumber.has(key)) {
+      // add || tp avoid NaN and pass check
+      mapValue.set(key, (mapValue.get(key) || 0) + item.value)
+      mapNumber.set(key, (mapNumber.get(key) || 0) + 1)
+    } else {
+      mapValue.set(key, item.value)
+      mapNumber.set(key, 1)
+    }
+  }
+
+  for (const [key, value] of mapValue) {
+    map.set(key, value / mapNumber.get(key)!)
+  }
+  return map
+}
+
+function convertFromMapAndAgg(
+  map: Map<string, number>
+): BackPressureRateInfo[] {
+  const result: BackPressureRateInfo[] = []
+  map.forEach((value, key) => {
+    const [fragmentId, downstreamFragmentId] = key.split("-").map(Number)
+    const backPressureRateInfo: BackPressureRateInfo = {
+      actorId: 0,
+      fragmentId,
+      downstreamFragmentId,
+      backPressureRate: value,
+    }
+    result.push(backPressureRateInfo)
+  })
+  return result
+}
+
+function convertToBackPressureMetrics(
+  bpRates: BackPressureRateInfo[]
+): BackPressuresMetrics {
+  const bpMetrics: BackPressuresMetrics = {
+    outputBufferBlockingDuration: [],
+  }
+  for (const item of bpRates) {
+    bpMetrics.outputBufferBlockingDuration.push({
+      metric: {
+        actorId: item.actorId.toString(),
+        fragmentId: item.fragmentId.toString(),
+        downstreamFragmentId: item.downstreamFragmentId.toString(),
+      },
+      sample: [
+        {
+          timestamp: Date.now(),
+          value: item.backPressureRate,
+        },
+      ],
+    })
+  }
+  return bpMetrics
+}
+
+export function calculateBPRate(
+  backPressureNew: BackPressureInfo[],
+  backPressureOld: BackPressureInfo[]
+): BackPressuresMetrics {
+  let mapNew = convertToMapAndAgg(backPressureNew)
+  let mapOld = convertToMapAndAgg(backPressureOld)
+  let result = new Map<string, number>()
+  mapNew.forEach((value, key) => {
+    if (mapOld.has(key)) {
+      result.set(
+        key,
+        // The *100 in end of the formular is to convert the BP rate to the value used in web UI drawing
+        ((value - (mapOld.get(key) || 0)) / ((INTERVAL / 1000) * 1000000000)) *
+          100
+      )
+    } else {
+      result.set(key, 0)
+    }
+  })
+
+  return convertToBackPressureMetrics(convertFromMapAndAgg(result))
+}
+
+export const BackPressureInfo = {
+  fromJSON: (object: any) => {
+    return {
+      actorId: isSet(object.actorId) ? Number(object.actorId) : 0,
+      fragmentId: isSet(object.fragmentId) ? Number(object.fragmentId) : 0,
+      downstreamFragmentId: isSet(object.downstreamFragmentId)
+        ? Number(object.downstreamFragmentId)
+        : 0,
+      value: isSet(object.value) ? Number(object.value) : 0,
+    }
+  },
+}
+
+// Get back pressure from meta node -> compute node
+export async function getBackPressureWithoutPrometheus() {
+  const response = await api.get("/metrics/back_pressures")
```

Comment:
Shall we move it to another path? It's mixed with the metrics fetched from Prometheus, which looks confusing to me.

https://github.com/risingwavelabs/risingwave/blob/ae7d7ba426818d43209be545ad72f979ce3dacc0/src/meta/src/dashboard/mod.rs#L423-L428

Reply:
Btw, `/metrics/actor/back_pressures` actually output **fragment-level** metrics instead of actor-level... 😂 Legacy issues

Reply:
Will solve the problem in a seperate PR